pallets/amm/src/lib.rs

```diff
@@ -465,3 +483,127 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+                    .checked_sub(lp_fees)
+                    .ok_or(ArithmeticError::Underflow)?;
+
+                // 3. Given the input amount amount_in left after fees, compute amount_out
+                // let amount_out = amount_in * supply_out / (supply_in + amount_in)
+                let amount_out = amount_in_after_all_fees
+                    .saturating_mul(supply_out)
+                    .checked_div(
+                        supply_in
+                            .checked_add(amount_in_after_all_fees)
+                            .ok_or(ArithmeticError::Overflow)?,
+                    )
+                    .ok_or(ArithmeticError::Underflow)?;
+
+                // 4. If `amount_out` is lower than `min_amount_out`, error
+                ensure!(
+                    amount_out >= minimum_amount_out && amount_in > 0,
+                    Error::<T, I>::InsufficientAmountOut
+                );
+
+                // 5. Update the `Pools` storage to track the `base_amount` and `quote_amount`
+                // variables (increase and decrease by `amount_in` and `amount_out`)
+                // increase liquidity_amount.base_amount by amount_in, unless inverted
+                if is_inverted {
+                    liquidity_amount.quote_amount = liquidity_amount
+                        .quote_amount
+                        .checked_add(amount_without_protocol_fees)
+                        .ok_or(ArithmeticError::Overflow)?;
+
+                    liquidity_amount.base_amount = liquidity_amount
+                        .base_amount
+                        .checked_sub(amount_out)
+                        .ok_or(ArithmeticError::Underflow)?;
+                } else {
+                    liquidity_amount.base_amount = liquidity_amount
+                        .base_amount
+                        .checked_add(amount_without_protocol_fees)
+                        .ok_or(ArithmeticError::Overflow)?;
+
+                    liquidity_amount.quote_amount = liquidity_amount
+                        .quote_amount
+                        .checked_sub(amount_out)
+                        .ok_or(ArithmeticError::Underflow)?;
+                }
+                *pool_liquidity_amount = Some(liquidity_amount);
+
+                // 6. Wire amount_in of the input token (identified by pair.0) from who to PalletId
+                T::Currency::transfer(
+                    pair.0,
+                    &who,
+                    &Self::account_id(),
+                    amount_without_protocol_fees,
+                )?;
+
+                // 7. Wire amount_out of the output token (identified by pair.1) to who from PalletId
+                T::Currency::transfer(pair.1, &Self::account_id(), &who, amount_out)?;
+
+                // 8. Wire protocol fees as needed (input token)
+                T::Currency::transfer(pair.0, &who, &T::ProtocolFeeReceiver::get(), protocol_fees)?;
```

Comment:
```suggestion
                T::Currency::transfer(base_asset, &who, &T::ProtocolFeeReceiver::get(), protocol_fees)?;
```

Reply:
same reasoning as the comment above.  

Earlier in the `trade` function we use the following if statement: 

```rust
// supply_in == liquidity_amount.base_amount unless inverted
let (supply_in, supply_out) = if is_inverted {
    (liquidity_amount.quote_amount, liquidity_amount.base_amount)
} else {
    (liquidity_amount.base_amount, liquidity_amount.quote_amount)
};
```

The above statement handles the case where `pair.0 != base_asset`. Therefore we'll need access the values from the tuple directly